### PR TITLE
feat: support variable when rewrite header in proxy rewrite plugin

### PR DESCRIPTION
**Pull request by monkeyDluffy6017**
### Description

Support for referencing the match result of the `regex_uri`  when rewrite header in the proxy-rewrite plugin, and the match result can be referenced as a variable by other plugins too.

Nginx: 
```
location ~* ^/aaa/(.*)$ {
    ...
    proxy_set_header HDR_TEST $1;
}

```

Apisix:
```
"plugins": {
    "proxy-rewrite": {
        "regex_uri": ["^/test/(.*)/(.*)/(.*)", "/$1_$2_$3"],
        "headers": {
            "X-Api:Version": "v2-$1"
        }
    }
},
```

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [x] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

File: apisix/core/utils.lua [hunk 1]
```diff
@@ -335,4 +335,45 @@ end
 _M.resolve_var = resolve_var
 
 
+local resolve_var_with_captures
+do
+    local _captures
+    -- escape is not supported very well, like there si a redundant '\' after escape "$1"
```
Comment (spacewander) [inline apisix/core/utils.lua:341]:
```suggestion
    -- escape is not supported very well, like there is a redundant '\' after escape "$1"
```

File: apisix/plugins/proxy-rewrite.lua [hunk 2]
```diff
@@ -278,15 +284,24 @@ function _M.rewrite(conf, ctx)
...
                         conf.regex_uri[2] .. " : " .. err
             core.log.error(msg)
             return 500, {message = msg}
         end
+
+        local m, err = re_match(upstream_uri, conf.regex_uri[1], "jo")
+        if not m then
+            if err then
+                core.log.error("match error in proxy-rewrite plugin, please check: ", err)
+                return
```
Comment (spacewander) [inline apisix/plugins/proxy-rewrite.lua:299]: Do we need to return a 503 for this error?

File: apisix/plugins/proxy-rewrite.lua [hunk 3]
```diff
@@ -278,15 +284,24 @@ function _M.rewrite(conf, ctx)
...
             local msg = "failed to substitute the uri " .. ctx.var.uri ..
                         " (" .. conf.regex_uri[1] .. ") with " ..
                         conf.regex_uri[2] .. " : " .. err
             core.log.error(msg)
             return 500, {message = msg}
         end
+
+        local m, err = re_match(upstream_uri, conf.regex_uri[1], "jo")
+        if not m then
+            if err then
```
Comment (spacewander) [inline apisix/plugins/proxy-rewrite.lua:297]: We can merge the two conditions above?

File: apisix/core/utils.lua [hunk 4]
```diff
@@ -335,4 +335,45 @@ end
...
+    local pat = [[ (?<! \\) \$ \{? (\d+) \}? ]]
+
+    local function resolve(m)
+        local v = _captures[tonumber(m[1])]
```
Comment (spacewander) [inline apisix/core/utils.lua:345]: Do we need to support capture group name?

Comment (leslie-tsang) [reply apisix/core/utils.lua:345]: The current implementation meets the needs, seems no need to support the capture group name. :)

File: apisix/plugins/proxy-rewrite.lua (hunk 2 above)
Comment (monkeyDluffy6017) [reply apisix/plugins/proxy-rewrite.lua:299]: I prefer 500

File: apisix/plugins/proxy-rewrite.lua (hunk 3 above)
Comment (monkeyDluffy6017) [reply apisix/plugins/proxy-rewrite.lua:297]: how? use the `re_match` first, then use the match result to replace?

Comment (leslie-tsang) [inline apisix/plugins/proxy-rewrite.lua:297]:
```suggestion
        if not m and err then
```
Would be better ? Seems no necessary to separate them.

Comment (spacewander) [reply apisix/plugins/proxy-rewrite.lua:297]: See https://github.com/apache/apisix/pull/9112#discussion_r1145538773

File: apisix/core/utils.lua (hunk 1 above)
Comment (monkeyDluffy6017) [reply apisix/core/utils.lua]: Done

File: apisix/plugins/proxy-rewrite.lua (hunk 3 above)
Comment (monkeyDluffy6017) [reply apisix/plugins/proxy-rewrite.lua]: Done

**Review by leslie-tsang — APPROVED** [review]

File: apisix/plugins/proxy-rewrite.lua [hunk 5]
```diff
@@ -41,6 +42,10 @@ local lrucache = core.lrucache.new({
     type = "plugin",
 })
 
+core.ctx.register_var("proxy_rewrite_regex_uri_captures", function(ctx)
+    return ctx.proxy_rewrite_regex_uri_captures
```
Comment (soulbird) [inline apisix/plugins/proxy-rewrite.lua:46]: Why need this?

Comment (soulbird) [reply apisix/plugins/proxy-rewrite.lua:46]: ok

**Review by soulbird — APPROVED** [review]

**Review by spacewander — APPROVED** [review]

**Pull request merged**